features/continue_as_new/updates_do_not_block_continue_as_new/feature.py

```diff
@@ -0,0 +1,103 @@
+processing a workflow task (WFT) at the time that the update is admitted by the server. To
+do this it must synchronize the workflow and client. This uses techniques that should
+never be used in real workflows. The synchronization must be threading-based as opposed to
+asyncio-based, since the point is to not allow the workflow to yield while it is waiting
+for notification from the client. In order for the workflow and client to share the same
+module namespace, we use UnsandboxedWorkflowRunner. But this means that the workflow and
+client code execute in the same thread. Therefore we create a new thread for the client
+code to execute in, so that the two can use thread-blocking waits on the shared
+threading.Event object.
+"""
+
+import asyncio
+import threading
+from datetime import timedelta
+from uuid import uuid4
+
+from temporalio import workflow
+from temporalio.client import WorkflowHandle
+from temporalio.worker import UnsandboxedWorkflowRunner, WorkerConfig
+
+from harness.python.feature import Runner, register_feature
+from harness.python.util import admitted_update_task
+
+# See docstring at top of file.
+first_run_wft_is_in_progress = threading.Event()
+update_has_been_admitted = threading.Event()
+
+
+@workflow.defn
+class Workflow:
+    def __init__(self):
+        self.received_update = False
+
+    @workflow.run
+    async def run(self) -> str:
+        """
+        Continue as new once, then return the current run ID.
+        """
+        if not first_run_wft_is_in_progress.is_set():
+            # Note: you should usually never block the thread in workflow code.
+            # See docstring at top of file.
+            first_run_wft_is_in_progress.set()
+            update_has_been_admitted.wait()
+
+        info = workflow.info()
+        if info.continued_run_id is not None:
+            # The update is probably delivered in the first post-CAN WFT, in which case
+            # the following wait_condition is not needed. However, correct behavior does
+            # not require this to be true.
+            await workflow.wait_condition(lambda: self.received_update)
+            return info.run_id
+
+        workflow.continue_as_new()
+
+    @workflow.update
+    async def update(self) -> str:
+        """Update handler that returns the current run ID"""
+        self.received_update = True
+        return workflow.info().run_id
+
+
+async def start(runner: Runner) -> WorkflowHandle:
+    return await runner.client.start_workflow(
+        Workflow.run,
+        id=str(uuid4()),
+        task_queue=runner.task_queue,
+        execution_timeout=timedelta(minutes=1),
+    )
+
+
+async def check_result(runner: Runner, handle: WorkflowHandle) -> None:
+    async def _check_result() -> None:
+        # See docstring at top of file.
+        # Cause an update to be admitted while the first WFT is in progress
+        first_run_wft_is_in_progress.wait()
```

Comment:
This is a thread-blocking call in asyncio. Can you use https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor? Would also request on the in-workflow `wait` but that makes a bit less sense since blocking the workflow event loop is the goal.

Reply:
> This is a thread-blocking call in asyncio.

Yes that's deliberate, could you read the docstring at the top of the file and let me know if you agree with the strategy?

Reply:
```suggestion
        asyncio.get_running_loop().run_in_executor(None, first_run_wft_is_in_progress.wait)
```

^ That is the safe way to make a thread-blocking call like this

Reply:
I believe the code is doing this correctly already. This entire coroutine is run in a new event loop in a new thread. See line 93 below:

```python
await asyncio.to_thread(lambda: asyncio.run(_check_result()))
```

Reply:
Ah, I see, then I think this can bring up a different issue: our client is not safe for running in separate event loops (because it is not safe for running in separate threads, and each event loop is effectively a thread)

Reply:
Ah, right. Modified to just do the thread-blocking synchronization itself in a new thread, using `asyncio.to_thread`:
```python
await asyncio.to_thread(first_run_wft_is_in_progress.wait)
```